### ISSUE TITLE
Improvements for GreaseMonkey support

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#3238]
Things which could be improved after #3040 is in:

- [ ] Maybe offer an install dialog when a `.user.js` file is opened, [like Chromium](https://www.chromium.org/developers/design-documents/user-scripts)
- [ ] Support [`@require`](https://wiki.greasespot.net/Third-Party_Libraries) (maybe without the downloading, but at least load those files if they exist?)
- [ ] Try stubbing [exportFunction](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Language_Bindings/Components.utils.exportFunction) and maybe set a more correct `scriptHandler` and `version` then for [GM_info](https://wiki.greasespot.net/GM_info)?
- [ ] Support `isPrivate` in [GM_info](https://wiki.greasespot.net/GM_info).
- [ ] Look into GreaseMonkey's sandboxing/security features
- [ ] Set up a [QtWebChannel](http://doc.qt.io/qt-5/qtwebchannel-index.html) so we can give scripts elevated permissions
    - [ ] Use some kind of database for [GM_getValue](https://wiki.greasespot.net/GM_getValue) et al so it's isolated from the page's localStorage
    - [ ] Use a [QNetworkAccessManager](http://doc.qt.io/qt-5/qnetworkaccessmanager.html) for [GM_xmlhttpRequest](https://wiki.greasespot.net/GM_xmlhttpRequest) so it can be cross-origin.
    - [ ] Maybe also something for [GM_openInTab](https://wiki.greasespot.net/GM_openInTab)

**Comment by jgkamat** [comment]
If anyone is thinking about using webchannels, [I have a branch](https://github.com/jgkamat/qutebrowser/tree/jay/webchannel) you can steal from where I started to work on this (to support follow_selected a while back) but I trashed it because it was not needed.

**Comment by toofar** [comment]
Also support regular expressions in `@include` and friends, we are currently only supporting globbing via `fnmatch`. zro in irc pointed out a good example script https://github.com/StylishThemes/GitHub-Dark-Script

**Comment by The-Compiler** [comment]
Known broken scripts:

- [Rotten Tomatoes Link On IMDb](https://github.com/Ede123/userscripts/blob/master/Rotten_Tomatoes_Link_On_IMDb.user.js) - needs cross-origin GM_XHR
- [OneeChan](https://github.com/Nebukazar/OneeChan) - seems to set `window.$` to its own (non-jQuery) thing which conflicts with the `window.$` of 4chan or 4chan-x (not sure which)

**Comment by The-Compiler** [comment]
OneeChan (and anything else defining global variables which collide with the page or other scripts) now works properly since #3793 - thanks @toofar!

**Comment by rnhmjoj** [comment]
I have found an issue with a script I made:
```js
// ==UserScript==
// @name         Kill subreddit style
// @description  Unload all subreddits custom stylesheets
// @author       rnhmjoj
// @include      https://old.reddit.com/r/*
// @version      1.0.0
// @run-at       document-start
// @grant        none
// ==/UserScript==

for(var link of document.getElementsByTagName("link")) {
  if(link.title == "applied_subreddit_stylesheet")
    link.disabled = true;
}
```

The script doesn't work because `document.getElementsByTagName("link")` returns an empty array when running from greasemonkey, pasting the code into the console however works fine.

**Comment by toofar** [comment]
That's just because it is running too early before those elements have been added. Try changing the `// @run-at       document-start` to `// @run-at       document-end`. (`document-idle` is even later but doesn't appear to be necessary here.)

**Comment by rnhmjoj** [comment]
Ah, thank you! it's working. This used to work with violentmonkey... is it just slower to load the scripts then?

**Comment by toofar** [comment]
Oh that's interesting, and annoying. QtWebEngine handles the events for us so maybe that gets access to earlier triggers than extensions do.

**Comment by The-Compiler** [comment]
The [GreaseMonkey wiki](https://wiki.greasespot.net/Metadata_Block#.40run-at) says that only `document-end` works in GreaseMonkey 4.x (the WebExtensions port) - maybe something similar is true for ViolentMonkey as well.

FWIW the [document.readyState docs](https://developer.mozilla.org/en-US/docs/Web/API/Document/readyState) say "We can now access the DOM elements." for the "interactive" state, which implies we can't for the "loading" state, which "document-start" corresponds to.

In other words, I think this is fine, and your script should use `document-end` :wink:

**Comment by slashome** [comment]
Don't know if I should be this here , but I made a script for Trello's plugin that add cost to task that display the total cost for a list. It works fine on Chrome but it seems to have a conflict with newrelic.js in qutebroser : 
```
Uncaught TypeError: Illegal invocation
    at Proxy.s (newrelic.js:664)
    at userscript:GM-Trello Cost Counter:245
    at userscript:GM-Trello Cost Counter:248
    at userscript:GM-Trello Cost Counter:253
```
The rewrelic.js line : 
```
            try {
              return s = t.apply(a, r)
            } catch (d) {
              throw u(e + "err", [r, a, d], c), d         <----
            } finally {
              u(e + "end", [r, a, s], c)
            }
```

My script link is : 
https://github.com/slashome/userscripts

Don't know if there is anything I can try to fix it :/

Thx again for this awesome browser !

**Comment by toofar** [comment]
Yeah that looks like they are shadowing the browser's `Proxy`
implementation. Since you are only modifying the DOM (as opposed to
talking to and trello JS API object) you should be able to run it in a
different JS world to isolate it from that. Add this line to the
metadata (see #4017 for more info):

```
// @qute-js-world jseval
```

**Comment by user202729** [comment]
* `@match https://stuff.com/*` also matches `view-source:https://stuff.com/`.
* Perhaps the script name should be edited so clicking the script name (in `console.log` messages in the inspector tools for example) opens it? When there are spaces or `-` in the name it doesn't work. (This might be a chromium bug fixed in some recent version. I'm not sure)

**Comment by amacfie** [comment]
Does `GM_openInTab` currently require
`content.javascript.can_open_tabs_automatically = True`?

**Comment by The-Compiler** [comment]
@amacfie I think so - why are you asking?

**Comment by amacfie** [comment]
@The-Compiler I was wondering why it wasn't working

**Comment by LiteracyFanatic** [comment]
It would be nice if there was a setting to enable/disable the loading of GreaseMonkey scripts. Sometimes I will encounter odd behavior on a page which I suspect may be related to one of my scripts and would like to quickly check if the page functions correctly without them. Right now I append `.disabled` to the end of each script and then run `:greasemonkey-reload` and `reload` as a workaround.

**Comment by The-Compiler** [comment]
Similarly to what @LiteracyFanatic already mentioned, I think it'd be great to have `:greasemonkey-{dis,en}able` commands which either disable/enable all scripts, or the script given to it (with completion of currently disabled/enabled scripts).

**Comment by Ninlives** [comment]
Any good news for cross-origin GM_xmlhttpRequest?

**Comment by toofar** [comment]
@Ninlives no good news sorry. None on the horizon either.
I think it came up last year but I don't even remember how the discussion went. "It" being how we could or couldn't provide a privileged (compared to the normal JS APIs) bridge between python and javascript that we were comfortable inflicting on all of the qutebrowser users. The risk being that javascript coming from places other than greasemonkey scripts could make use of the same channel. It looks like a tough problem still and we have plenty of other things that require more urgent attention.

**Comment by cortsf** [comment]
Just leaving this comment to upvote implementing `GM_getResourceText`. It was mentionted that nobody else asked for it on #3975 so I'm just raising one more hand.

Do you think there is any workaround in the meantime? For example, reading a local file with python in config.py, and setting a variable which can be read from the greasemonkey script?

Thanks

**Comment by The-Compiler** [comment]
The obvious workaround is to include your resource in your JS file. Other than that, no. There's no way for Greasemonkey to access Python stuff, see the comment just above yours.

**Comment by cortsf** [comment]
Thanks. Definitely not a big problem to keep doing it, but my particular use case is actually to avoid doing that, to avoid having to maintain multiple copies of the same text in multiple places. Can live with it.
